File: ios-bookmind/Sources/ReaderCore/TXTParser.swift
import Foundation
import SharedModels

public final class TXTParser: BookParsing {
    public init() {}

    private static let chapterMarkers: [NSRegularExpression] = {
        let patterns = [
            #"^\s*Chapter\s+\d+.*$"#,
            #"^\s*CHAPTER\s+[IVXLC\d]+.*$"#,
            #"^\s*Глава\s+\d+.*$"#,
            #"^\s*ГЛАВА\s+[IVXLC\d]+.*$"#,
            #"^\s*\d+\.\s+\p{Lu}.*$"#
        ]
        return patterns.compactMap {
            try? NSRegularExpression(pattern: $0, options: [.anchorsMatchLines])
        }
    }()

    public func parseChapters(for book: Book) async throws -> [Chapter] {
        let text = try readText(at: book.fileURL)
        let splits = splitChapters(text: text)

        if splits.isEmpty {
            // Single-chapter fallback
            return [
                Chapter(
                    id: ChapterID("\(book.id.rawValue)#0"),
                    bookID: book.id,
                    index: 0,
                    title: book.title,
                    contentRef: book.fileURL
                )
            ]
        }

        return splits.enumerated().map { idx, slice in
            Chapter(
                id: ChapterID("\(book.id.rawValue)#\(idx)"),
                bookID: book.id,
                index: idx,
                title: slice.title,
                contentRef: book.fileURL
            )
        }
    }

    public func rawText(for chapter: Chapter, in book: Book) async throws -> String {
        let full = try readText(at: book.fileURL)
        let splits = splitChapters(text: full)
        if splits.isEmpty { return full }
        guard chapter.index >= 0, chapter.index < splits.count else { return "" }
        return splits[chapter.index].body
    }

    // MARK: - Internal

    struct ChapterSlice {
        let title: String?
        let body: String
    }

    func splitChapters(text: String) -> [ChapterSlice] {
        let nsText = text as NSString
        var matches: [(range: NSRange, title: String)] = []
        for regex in Self.chapterMarkers {
            let found = regex.matches(in: text, range: NSRange(location: 0, length: nsText.length))
            for m in found {
                matches.append((m.range, nsText.substring(with: m.range).trimmingCharacters(in: .whitespacesAndNewlines)))
            }
        }
        matches.sort { $0.range.location < $1.range.location }

        guard !matches.isEmpty else { return [] }

        var slices: [ChapterSlice] = []
        for (i, m) in matches.enumerated() {
            let bodyStart = m.range.location + m.range.length
            let bodyEnd = (i + 1 < matches.count) ? matches[i + 1].range.location : nsText.length
            guard bodyEnd > bodyStart else { continue }
            let body = nsText.substring(with: NSRange(location: bodyStart, length: bodyEnd - bodyStart))
                .trimmingCharacters(in: .whitespacesAndNewlines)
            slices.append(ChapterSlice(title: m.title, body: body))
        }
        return slices
    }

    private func readText(at url: URL) throws -> String {
        guard FileManager.default.isReadableFile(atPath: url.path) else {
            throw ReaderError.fileNotReadable(url)
        }
        let encodings: [String.Encoding] = [.utf8, .utf16, .isoLatin1, .windowsCP1251]
        for enc in encodings {
            if let s = try? String(contentsOf: url, encoding: enc) { return s }
        }
        throw ReaderError.parsingFailed("Could not decode text file")
    }
}

File: android-bookmind/app/src/main/java/com/bookmind/core/parser/TxtParser.kt
package com.bookmind.core.parser

import android.content.Context
import android.net.Uri
import com.bookmind.core.model.Book
import com.bookmind.core.model.Chapter
import com.bookmind.core.model.ChapterID
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.nio.charset.Charset
import javax.inject.Inject

/** = iOS `TXTParser`. Heuristically splits plain text into chapters. */
class TxtParser @Inject constructor(
    @ApplicationContext private val context: Context
) : BookParsing {

    data class ChapterSlice(val title: String?, val body: String)

    override suspend fun parseChapters(book: Book): List<Chapter> = withContext(Dispatchers.IO) {
        val text = readText(book.fileUri)
        val splits = splitChapters(text)

        if (splits.isEmpty()) {
            // Single-chapter fallback.
            return@withContext listOf(
                Chapter(
                    id = ChapterID("${book.id.raw}#0"),
                    bookID = book.id,
                    index = 0,
                    title = book.title,
                    contentRef = book.fileUri
                )
            )
        }

        splits.mapIndexed { idx, slice ->
            Chapter(
                id = ChapterID("${book.id.raw}#$idx"),
                bookID = book.id,
                index = idx,
                title = slice.title,
                contentRef = book.fileUri
            )
        }
    }

    override suspend fun rawText(chapter: Chapter, book: Book): String = withContext(Dispatchers.IO) {
        val full = readText(book.fileUri)
        val splits = splitChapters(full)
        if (splits.isEmpty()) return@withContext full
        if (chapter.index !in splits.indices) return@withContext ""
        splits[chapter.index].body
    }

    // MARK: - Internal

    fun splitChapters(text: String): List<ChapterSlice> {
        val matches = chapterMarkers
            .flatMap { regex -> regex.findAll(text).map { it.range.first to it.value.trim() } }
            .sortedBy { it.first }

        if (matches.isEmpty()) return emptyList()

        val slices = mutableListOf<ChapterSlice>()
        for (i in matches.indices) {
            val (start, title) = matches[i]
            val bodyStart = start + title.length
            val bodyEnd = if (i + 1 < matches.size) matches[i + 1].first else text.length
            if (bodyEnd <= bodyStart) continue
            val body = text.substring(bodyStart, bodyEnd).trim()
            slices.add(ChapterSlice(title, body))
        }
        return slices
    }

    private fun readText(uri: String): String {
        val bytes = context.contentResolver.openInputStream(Uri.parse(uri))?.use { it.readBytes() }
            ?: throw ReaderError.FileNotReadable(uri)
        val encodings = listOf(Charsets.UTF_8, Charsets.UTF_16, Charsets.ISO_8859_1, windows1251)
        for (enc in encodings) {
            runCatching { decodeStrict(bytes, enc) }.getOrNull()?.let { return it }
        }
        throw ReaderError.ParsingFailed("Could not decode text file")
    }

    private fun decodeStrict(bytes: ByteArray, charset: Charset): String {
        val decoder = charset.newDecoder() // throws on malformed input
        return decoder.decode(java.nio.ByteBuffer.wrap(bytes)).toString()
    }

    companion object {
        private val windows1251: Charset = Charset.forName("windows-1251")

        private val chapterMarkers: List<Regex> = listOf(
            Regex("""^\s*Chapter\s+\d+.*$""", RegexOption.MULTILINE),
            Regex("""^\s*CHAPTER\s+[IVXLC\d]+.*$""", RegexOption.MULTILINE),
            Regex("""^\s*Глава\s+\d+.*$""", RegexOption.MULTILINE),
            Regex("""^\s*ГЛАВА\s+[IVXLC\d]+.*$""", RegexOption.MULTILINE),
            Regex("""^\s*\d+\.\s+\p{Lu}.*$""", RegexOption.MULTILINE)
        )
    }
}
